xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go

```diff
@@ -1370,46 +1180,122 @@ func (s) TestAggregateCluster_ErrorsFromEDSAndDNS(t *testing.T) {
-	if err != nil {
-		t.Fatalf("Failed to create xDS client: %v", err)
+	// Create xDS client, configure cds_experimental LB policy with a manual
+	// resolver, and dial the test backends.
+	cc, cleanup := setupAndDial(t, bootstrapContents)
+	defer cleanup()
+
+	// Make an RPC with a short deadline. We expect this RPC to not succeed
+	// because the EDS resource came back with no endpoints, and we are yet to
+	// push an update through the DNS resolver.
+	client := testgrpc.NewTestServiceClient(cc)
+	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+	if _, err := client.EmptyCall(sCtx, &testpb.Empty{}); status.Code(err) != codes.DeadlineExceeded {
+		t.Fatalf("EmptyCall() code %s, want %s", status.Code(err), codes.DeadlineExceeded)
 	}
-	defer close()
 
-	// Create a manual resolver and push a service config specifying the use of
-	// the cds LB policy as the top-level LB policy, and a corresponding config
-	// with a single cluster.
-	r := manual.NewBuilderWithScheme("whatever")
-	jsonSC := fmt.Sprintf(`{
-			"loadBalancingConfig":[{
-				"cds_experimental":{
-					"cluster": "%s"
-				}
-			}]
-		}`, clusterName)
-	scpr := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(jsonSC)
-	r.InitialState(xdsclient.SetClient(resolver.State{ServiceConfig: scpr}, xdsClient))
+	// Ensure that the DNS resolver is started for the expected target.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for DNS resolver to be started")
+	case target := <-dnsTargetCh:
+		got, want := target.Endpoint(), fmt.Sprintf("%s:%d", dnsHostName, dnsPort)
+		if got != want {
+			t.Fatalf("DNS resolution started for target %q, want %q", got, want)
+		}
+	}
 
-	// Create a ClientConn.
-	cc, err := grpc.Dial(r.Scheme()+":///test.service", grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(r))
-	if err != nil {
-		t.Fatalf("Failed to dial local test server: %v", err)
+	// Update DNS resolver with test backend addresses.
+	dnsR.UpdateState(resolver.State{Addresses: addrs})
+
+	// Ensure that RPCs start getting routed to the first backend since the
+	// child policy for a LOGICAL_DNS cluster is pick_first by default.
+	for ; ctx.Err() == nil; <-time.After(defaultTestShortTimeout) {
+		peer := &peer.Peer{}
+		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(peer)); err != nil {
+			t.Logf("EmptyCall() failed: %v", err)
+			continue
+		}
+		if peer.Addr.String() == addrs[0].Addr {
+			break
+		}
+	}
+	if ctx.Err() != nil {
+		t.Fatalf("Timeout when waiting for RPCs to be routed to backend %q in the DNS cluster", addrs[0].Addr)
+	}
+
+	// Push an error from the DNS resolver as well.
+	dnsErr := fmt.Errorf("DNS error")
+	dnsR.ReportError(dnsErr)
+
+	// Ensure that RPCs continue to succeed for the next one second.
+	for end := time.Now().Add(time.Second); time.Now().Before(end); <-time.After(defaultTestShortTimeout) {
```

Comment:
I don't feel strongly about this, but I feel I've seen patterns where you time.Sleep for a millisecond at the end of for and for 0 - 1000 or spawn a Timer and wait until done with time.Sleep at the end. I've never seen this type of for before. I'm curious as to whether you know the style guide specification for the best practices for situations like this (sort of polling, whereas in this case the event you poll until is the second passing).

Reply:
I haven't seen any guidance in the style guide about this. But to answer your question, I think we arrived at this pattern recently after a few iterations of doing the same thing in different ways. I had initially gotten some comments from Doug which probably got me to half-way here, and then I think Arvind came up with this simplification.

Things that I can recollect that we used to before:
```golang
for i := 0; i < N; i++ {
  // Do stuff
  // time.Sleep(10*time.Millisecond)
}
```

```golang
// Create context with appropriate deadline
for ; ctx.Err() == nil; <-time.After(10*Millisecond) {
  // Do stuff
}
```

Reply:
Yeah those are the two I've seen as well. Sounds good.